pkg/storage/storage_tests/storage_test.go

```diff
@@ -0,0 +1,71 @@
+package storagetest
+
+import (
+	"testing"
+
+	"github.com/gobuffalo/suite"
+	"github.com/gomods/athens/pkg/config/env"
+	"github.com/gomods/athens/pkg/storage"
+	"github.com/gomods/athens/pkg/storage/fs"
+	"github.com/gomods/athens/pkg/storage/mem"
+	"github.com/gomods/athens/pkg/storage/minio"
+	"github.com/gomods/athens/pkg/storage/mongo"
+	"github.com/gomods/athens/pkg/storage/rdbms"
+	"github.com/spf13/afero"
+)
+
+type StorageTests struct {
+	*suite.Model
+	storages []storage.Backend
+}
+
+func (d *StorageTests) SetupTest() {
+	ra := d.Require()
+
+	// fs
+	memFs := afero.NewOsFs()
+	r, err := afero.TempDir(memFs, "", "athens-fs-storage-tests")
+	ra.NoError(err)
+
+	fsStore, err := fs.NewStorage(r, memFs)
+	ra.NoError(err)
+
+	d.storages = append(d.storages, fsStore)
+
+	// mem
+	memStore, err := mem.NewStorage()
+	ra.NoError(err)
+
+	d.storages = append(d.storages, memStore)
+
+	// minio
+	endpoint := "127.0.0.1:9000"
+	bucketName := "gomods"
+	accessKeyID := "minio"
+	secretAccessKey := "minio123"
+	minioStorage, err := minio.NewStorage(endpoint, accessKeyID, secretAccessKey, bucketName, false)
+	ra.NoError(err)
+
+	d.storages = append(d.storages, minioStorage)
+
+	// mongo
+	muri, err := env.MongoURI()
+	ra.NoError(err)
+
+	mongoStore := mongo.NewStorage(muri)
+	ra.NotNil(mongoStore)
+	ra.NoError(mongoStore.Connect())
+
+	d.storages = append(d.storages, mongoStore)
+
+	// // rdbms
+	conn := d.DB
+	rdbmsStore := rdbms.NewRDBMSStorageWithConn(conn)
+	d.Model.SetupTest()
+
+	d.storages = append(d.storages, rdbmsStore)
```

Comment:
@michalpristas how do you feel about inverting this test setup, so that the individual storage driver indicates how to create itself, and then the common tests exercise common test functionality. Here's a skeleton:

```go
type StorageTests struct {
    *suite.Model
    CreateStorage func() (storage.Storage, func(), error) // the individual storage driver sets this
    storage storage.Storage
    storageCloser func()
}

func (s StorageTests) SetupTest() {
    ra := d.Require()
    stg, close, err := s.CreateStorage()
    ra.NoError(err)
    s.storage = stg
    s.storageCloser = close
}

func (s StorageTests) TearDownTest() {
    s.storageCloser()
}

func (s StorageTests) TestGetSaveListRoundTrip() {
    // .....
}
```

Reply:
make sense, on the note of why not to move it. circular dependency when specific storage is using structs defined in storage root. and then test in storage root using specific storage. so i went with this

Reply:
I think there should be no tests in storage root that test functionality from sub packages. Those tests should be in each package.

For example in my WIP for gcp I have a new test for both modules and versions not being found.

Reply:
i think that in some cases it is good to have test like this. 
for example, this not found err. we rely on this error in our logic. so it is ok to have that in common place. 
specific things needs to be tested in each of them but not found or basic save-get roundtrip can be extracted

Reply:
Ok cool, and we can eventually move all the general implementation tests out into a suite of sorts?

Reply:
i forgot about one little detail, structs inside one test package cannot be used in another one. which results in as much complicated code as it is now so i will keep this as is

Reply:
@michalpristas RE the separate test package: that totally makes sense what you did.

regarding the idea of "inverting" the tests, I didn't understand what you meant in https://github.com/gomods/athens/pull/215#discussion_r200892945 - do you think it's a bad idea?